node-wasm/src/ports.rs

```diff
@@ -0,0 +1,306 @@
+
+use crate::commands::{NodeCommand, WorkerResponse};
+use crate::error::{Context, Error, Result};
+
+// Instead of supporting communication with just `MessagePort`, allow using any object which
+// provides compatible interface, eg. `Worker`
+#[wasm_bindgen]
+extern "C" {
+    pub type MessagePortLike;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message_with_transferable(
+        this: &MessagePortLike,
+        message: &JsValue,
+        transferable: &JsValue,
+    ) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(port: MessagePort) -> Self {
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
+        forward_messages_to: mpsc::UnboundedSender<(
+            ClientId,
+            Result<NodeCommand, TypedMessagePortError>,
+        )>,
+        forward_connects_to: mpsc::UnboundedSender<JsValue>,
+    ) -> Result<Self> {
+        let onmessage = Closure::new(move |ev: MessageEvent| {
+            let message_tx = forward_messages_to.clone();
+            let port_tx = forward_connects_to.clone();
+            spawn_local(async move {
+                let message: Result<NodeCommand, _> =
+                    from_value(ev.data()).map_err(TypedMessagePortError::FailedToConvertValue);
+
+                let ports = ev.ports();
+                if Array::is_array(&ports) {
+                    let port = ports.get(0);
+                    if !port.is_undefined() {
+                        if let Err(e) = port_tx.send(port) {
+                            error!("port forwarding channel closed, shouldn't happen: {e}");
+                        }
+                    }
+                }
+
+                if let Err(e) = message_tx.send((id, message)) {
+                    error!("message forwarding channel closed, shouldn't happen: {e}");
+                }
+            })
+        });
+
+        let port = prepare_message_port(object, &onmessage)
+            .context("failed to setup port for ClientConnection")?;
+
+        Ok(ClientConnection {
+            port,
+            _onmessage: onmessage,
+        })
+    }
+
+    fn send(&self, message: &WorkerResponse) -> Result<()> {
+        let serializer = Serializer::json_compatible();
+        let message_value = message
+            .serialize(&serializer)
+            .context("could not serialise message")?;
+        self.port
+            .post_message(&message_value)
+            .context("could not send command to worker")?;
+        Ok(())
+    }
+}
+
+pub struct RequestServer {
```

Comment:
I'm wondering about renaming `RequestServer` and `RequestResponse` to `WorkerServer` and `WorkerClient` respectively. initially I had a bit troubles to understand which part is for what. wdyt?